src/mjolnir.py

```diff
@@ -0,0 +1,220 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the Mjolnir class to represent the Mjolnir plugin for Synapse."""
+
+import logging
+import typing
+from secrets import token_hex
+
+import ops
+from ops.jujuversion import JujuVersion
+
+import actions
+import synapse
+from charm_state import CharmState
+from constants import (
+    MJOLNIR_MANAGEMENT_ROOM,
+    MJOLNIR_MEMBERSHIP_ROOM,
+    MJOLNIR_SERVICE_NAME,
+    PEER_RELATION_NAME,
+    SECRET_ID,
+    SECRET_KEY,
+    SYNAPSE_CONTAINER_NAME,
+)
+from user import User
+
+logger = logging.getLogger(__name__)
+
+MJOLNIR_USER = "mjolnir"
+
+
+class Mjolnir(ops.Object):  # pylint: disable=too-few-public-methods
+    """A class representing the Mjolnir plugin for Synapse application."""
+
+    def __init__(self, charm: ops.CharmBase, charm_state: CharmState):
+        """Initialize a new instance of the Mjolnir class.
+
+        Args:
+            charm: The charm object that the Mjolnir instance belongs to.
+            charm_state: Instance of CharmState.
+        """
+        super().__init__(charm, "mjolnir")
+        self._charm = charm
+        self._charm_state = charm_state
+        self.framework.observe(charm.on.collect_unit_status, self._on_collect_status)
```

Comment:
Should this be done on the charm rather than here?

Reply:
Only Mjolnir uses the collect_unit_status in order to check when the moderator room, a requirement for configuring the plugin, is created. This was used to avoid using defer in config-changed event. This was discussed on our MM channel, I can DM you the link to the thread :)

Reply:
Apologies I didn't get back on this, I think this is related to the conversation above so we can close this one for now and discuss there: https://github.com/canonical/synapse-operator/pull/30#discussion_r1324021010